lib/IRGen/LoadableByAddress.cpp

```diff
@@ -0,0 +1,1538 @@
+  for (auto *loadInstr : loadInstrsOfFunc) {
+    SILBuilder loadBuilder(loadInstr);
+    auto *newInstr =
+        loadBuilder.createLoad(loadInstr->getLoc(), loadInstr->getOperand(),
+                               loadInstr->getOwnershipQualifier());
+    loadInstr->replaceAllUsesWith(newInstr);
+    loadInstr->getParent()->erase(loadInstr);
+  }
+}
+
+void LoadableByAddress::recreateConvInstrs() {
+  for (auto *convInstr : conversionInstrs) {
+    IRGenModule *currIRMod =
+        getIRGenModule()->IRGen.getGenModule(convInstr->getFunction());
+    SILType currSILType = convInstr->getType();
+    CanType currCanType = currSILType.getSwiftRValueType();
+    SILFunctionType *currSILFunctionType =
+        dyn_cast<SILFunctionType>(currCanType.getPointer());
+    Lowering::GenericContextScope GenericScope(
+        getModule()->Types,
+        CanSILFunctionType(currSILFunctionType)->getGenericSignature());
+    if (!currSILFunctionType) {
+      llvm_unreachable("unsupported type");
+    }
+    SILType newType = getNewSILFunctionType(currSILFunctionType, *currIRMod);
+    SILBuilder convBuilder(convInstr);
+    SILInstruction *newInstr = nullptr;
+    switch (convInstr->getKind()) {
+    case ValueKind::ThinToThickFunctionInst: {
+      ThinToThickFunctionInst *instr =
+          dyn_cast<ThinToThickFunctionInst>(convInstr);
+      assert(instr && "Unexpected conversion instruction");
+      newInstr = convBuilder.createThinToThickFunction(
+          instr->getLoc(), instr->getOperand(), newType);
+      break;
+    }
+    case ValueKind::ConvertFunctionInst: {
+      ConvertFunctionInst *instr = dyn_cast<ConvertFunctionInst>(convInstr);
+      assert(instr && "Unexpected conversion instruction");
+      newInstr = convBuilder.createConvertFunction(
+          instr->getLoc(), instr->getOperand(), newType);
+      break;
+    }
+    default:
+      llvm_unreachable("Unexected conversion instruction");
+    }
+    convInstr->replaceAllUsesWith(newInstr);
+    convInstr->getParent()->erase(convInstr);
+  }
+}
+
+void LoadableByAddress::updateLoweredTypes(SILFunction *F) {
+  IRGenModule *currIRMod = getIRGenModule()->IRGen.getGenModule(F);
+  CanSILFunctionType funcType = F->getLoweredFunctionType();
+  Lowering::GenericContextScope GenericScope(getModule()->Types,
+                                             funcType->getGenericSignature());
+  SmallVector<SILParameterInfo, 4> newFuncArgTys =
+      getNewArgTys(F->getLoweredFunctionType()->getParameters(), *currIRMod);
+  SILFunctionType *OrigFTI = F->getLoweredFunctionType();
+  auto NewTy = SILFunctionType::get(
+      OrigFTI->getGenericSignature(), OrigFTI->getExtInfo(),
+      OrigFTI->getCalleeConvention(), newFuncArgTys, OrigFTI->getResults(),
+      OrigFTI->getOptionalErrorResult(), getModule()->getASTContext());
+  F->rewriteLoweredTypeUnsafe(NewTy);
+}
+
+/// The entry point to this function transformation.
+void LoadableByAddress::run() {
```

Comment:
It is called only once. But you are right: If it were called twice then we need to clear the state, I will clear the state at the end of `::run()` just to be safe.